Lib9c/Action/Coupons/RedeemCoupon.cs

```diff
@@ -0,0 +1,93 @@
+#nullable enable
+using System;
+using System.Collections.Immutable;
+using Bencodex.Types;
+using Libplanet;
+using Libplanet.Action;
+using Nekoyume.Model.Coupons;
+using Nekoyume.Model.Item;
+using Nekoyume.Model.State;
+using static Lib9c.SerializeKeys;
+
+namespace Nekoyume.Action.Coupons
+{
+    [Serializable]
+    [ActionType("redeem_coupon")]
+    public sealed class RedeemCoupon : GameAction
+    {
+        public Guid CouponId { get; private set; }
+        public Address AvatarAddress { get; private set; }
+
+        public RedeemCoupon(Guid couponId, Address avatarAddress)
+        {
+            CouponId = couponId;
+            AvatarAddress = avatarAddress;
+        }
+
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            var states = context.PreviousStates;
+            var inventoryAddress = AvatarAddress.Derive(LegacyInventoryKey);
+            var worldInformationAddress = AvatarAddress.Derive(LegacyWorldInformationKey);
+            var questListAddress = AvatarAddress.Derive(LegacyQuestListKey);
+            if (context.Rehearsal)
+            {
+                return states
+                    .SetState(AvatarAddress, MarkChanged)
+                    .SetState(inventoryAddress, MarkChanged)
+                    .SetState(worldInformationAddress, MarkChanged)
+                    .SetState(questListAddress, MarkChanged)
+                    .SetCouponWallet(
+                        context.Signer,
+                        ImmutableDictionary.Create<Guid, Coupon>(),
+                        rehearsal: true);
+            }
+
+            if (!states.TryGetAvatarStateV2(
+                    context.Signer,
+                    AvatarAddress,
+                    out AvatarState avatarState,
+                    out _))
```

Comment:
```suggestion
                    out bool migrationRequired))
```
 Can you take the `migrationRequired` param and use it like below?

```cs
if (migrationRequired)
{
    states = states
        .SetState(worldInformationAddress, avatarState.worldInformation.Serialize())
        .SetState(questListAddress, avatarState.questList.Serialize())
}

return states
    .SetState(AvatarAddress, avatarState.SerializeV2())
    .SetState(inventoryAddress, avatarState.inventory.Serialize())
    .SetCouponWallet(context.Signer, wallet);
```

Above is for optimization of action execution.🙂

Reply:
Fixed: 975aeebc466ac6a562a7b9440ccaf390679e925d.